IPIP/0504-provider-query-parameter.md

```diff
@@ -0,0 +1,242 @@
+
+The base format is:
+
+```sh!
+[ipfs://<CID> | https://<gateway>/ipfs/<CID> | https://<CID>.ipfs.<gateway> ]?[provider=<multiaddr1>&provider=<multiaddr22>&...]
+```
+
+### 🧠 Parsing
+
+The CID is the core of a Provider-Hinted URI. Clients MUST extract the CID before evaluating any hints. The format is designed to be compatible with current IPFS like URIs, while explicitly defining how to locate the CID and interpret `provider` query parameters.
+
+#### CID Extraction Rules
+
+To ensure consistent parsing, clients MUST extract the CID using the following precedence rules:
+
+**1. Multi-dotted Origin Format**
+
+If the CID is encoded as a subdomain label (e.g., `https://<CID>.ipfs.<gateway>`):
+
+- The CID MUST be the left-most label.
+- The label immediately following MUST be `ipfs`.
+- Any path MUST NOT also include a CID.
+
+**Example:**  
+✅ `https://bafy...ipfs.dweb.link`  
+❌ `https://bafy...ipfs.dweb.link/ipfs/bafybogus` (ambiguous; reject)
+
+**2. Path-Based Format**
+
+If the CID is encoded in the path (e.g., `https://<gateway>/ipfs/<CID>`):
+
+- The path MUST match the pattern `/ipfs/<CID>`, where `<CID>` is a valid content identifier.
+
+**Example:**  
+✅ `https://gateway.io/ipfs/bafy...`  
+❌ `https://gateway.io/bafy...` (no `/ipfs/` marker)
+
+**3. ipfs:// Scheme Format**
+
+- The CID MUST immediately follow the scheme delimiter: `ipfs://<CID>`.
+- Additional path/query components MAY follow.
+
+**Example:**  
+✅ `ipfs://bafy...`
+
+**4. Conflict Resolution**
+
+If a CID is present in both a multi-dotted origin and in the path (even if they match), the URI MUST be rejected as ambiguous.
+
+---
+
+#### Query Parameter: `provider`
+
+- Name: `provider`
+- Type: URI Query Parameter (repeating allowed)
+- Value: Multiaddr string (`?provider=multiaddr`).
+- Interpretation: Optional hint for how to fetch and locate the content identified by the CID
+
+#### Query Parsing (`provider` Parameters)
+
+Once a CID has been successfully extracted, clients MAY parse `provider` parameters from the query string. Each `provider` value represents a provider hint, encoded as a multiaddr string.
+
+**1. Parsing Rules**
+
+- The `provider` query parameter MAY appear multiple times.
+- Each `provider` parameter MUST be treated as an independent, optional provider hint.
+- Clients MAY ignore hints with invalid multiaddrs.
+
+**2. Evaluate hints**
+
+- Clients MAY:
+  - Ignore all `provider` parameters (if unsupported).
+  - Evaluate hints in order of appearance (left-to-right).
+  - Evaluate hints in parallel.
+  - Apply their own prioritization or fallback strategies. If all hints fail, clients SHOULD fall back to default discovery strategies (e.g., DHT/IPNI), if available. Or even rely on discovery strategies in parallel.
+
+Note that the `multiaddr` string should point to the `origin` server where given CID is provided, and not include the actual CID in the Hint multiaddr as a subdomain/path.
```

Comment:
Are CIDs valid in multiaddr at all (besides Peer ID)? might be worth linking to multiaddr spec: https://github.com/libp2p/specs/blob/master/addressing/README.md#multiaddr-in-libp2p